python/pyarrow/src/arrow/python/deserialize.cc

```diff
@@ -88,8 +88,19 @@ Status DeserializeDict(PyObject* context, const Array& array, int64_t start_idx,
     // The latter two steal references whereas PyDict_SetItem does not. So we need
     // to make sure the reference count is decremented by letting the OwnedRef
     // go out of scope at the end.
-    int ret = PyDict_SetItem(result.obj(), PyList_GET_ITEM(keys.obj(), i - start_idx),
-                             PyList_GET_ITEM(vals.obj(), i - start_idx));
+    PyObject *key, *val;
```

Comment:
What is the status of the deserialization code already? I see `DeserializeObject` is not being called anywhere, should it be entirely removed from the codebase? @jorisvandenbossche

Reply:
We removed the cython bindings (after a deprecation process) a while ago: https://github.com/apache/arrow/pull/34926. But it seems we forgot to remove the C++ code as well (or at least it is not clear why that was not done). 

Will open a separate issue for this.

Reply:
-> https://github.com/apache/arrow/issues/43587